### ISSUE TITLE
Gradient doesn't exist/ becomes None for conv_transpose layers

### ISSUE THREAD

**Issue by Dieblitzen** [tensorflow/tensorflow#34559]
**System information**
- Custom code
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOS Catalina
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: No
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.0
- Python version: 3.6.8
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source):N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Describe the current behavior**
I get a warning 
`WARNING:tensorflow:Gradients do not exist for variables ['mrf__block_2/conv2d_transpose_5/kernel:0', 'mrf__block_2/conv2d_transpose_5/bias:0'] when minimizing the loss.` 
when trying to backprop loss for a specific layer in my model. The gradients keep turning out to be None regardless of optimizer.

**Describe the expected behavior**
The gradients should exist for these layers.

**Code to reproduce the issue**
This is the model in question:
```
import tensorflow as tf
import tensorflow.keras.layers as layers
from tensorflow.keras import Model

class MRF_Block(Model):
  def __init__(self, out_channels, num_inputs, use_deconv=True, 
               kernel_size=(3,3), strides=(1,1), padding='same'):
    super(MRF_Block, self).__init__()

    # Add the conv/deconv layers.
    for i in range(num_inputs):
      # Convolve input tensors to output tensors of same channel depth (# of channels)
      conv_layer = layers.Conv2D(out_channels, kernel_size, strides=strides, padding=padding)
      setattr(self, f'conv_{i}', conv_layer)

      # Deconv input tensors. Input tensor at index i will be upsampled by factor 2^(n_inputs - (i+1))
      # Eg: if input tensors are [t1, t2, t3], t1 will be upsampled x 4, t2 upsampled x 2, t3 not upsampled.
      if use_deconv:
        deconv_layer = layers.Conv2DTranspose(out_channels, kernel_size, padding='same',
                                              strides=(2**(num_inputs-(i+1)), 2**(num_inputs-(i+1))))
        setattr(self, f'deconv_{i}', deconv_layer)

  ## inputs: a list of tensors being inputted to the block.
  def call(self, inputs):

    # Make #channels the same for input tensors (smallest depth)
    convolved = []
    for i, t in enumerate(inputs):
      conv = getattr(self, f'conv_{i}')
      convolved.append(conv(t))

    # Upsample to largest (h, w) resolution
    largest_res = max(inputs, key=lambda t: t.shape[1] * t.shape[2])
    resized = []
    for i, t in enumerate(convolved):
      deconv = getattr(self, f'deconv_{i}', lambda t: t)
      up_sampled = deconv(t)
      
      # Use bilinear resizing if resolution of optional conv_transpose doesn't match
      if up_sampled.shape[1:3] != largest_res.shape[1:3]:
        up_sampled = tf.image.resize(t, size=largest_res.shape[1:3])
    
      resized.append(up_sampled)

    # Fuse by summing
    return sum(resized)
```

And then running this code is what results in the error:
```
t = tf.random.normal((2, 224, 224, 3))
optimizer = tf.keras.optimizers.SGD()

mrf = MRF_Block(12, 2)
with tf.GradientTape() as tape: 
    preds = mrf([t, t]) 
    loss = tf.keras.losses.BinaryCrossentropy(from_logits=True)(tf.random.normal((2, 224, 224, 12)), preds) 
gradients = tape.gradient(loss, mrf.trainable_variables)
optimizer.apply_gradients(zip(gradients, mrf.trainable_variables))  
```
Changing SGD() to Adam() yields the same warning. Is it a problem with how the model is defined? Repeatedly running the above code yields the same warning (with/without reinitialising the optimizer).

**Other info / logs**
None at the moment.

**Comment by gowthamkpr** [comment]
@Dieblitzen Can you please provide a minimal reproducible code in colab as I cannot reproduce this issue. Thanks!

**Comment by Dieblitzen** [comment]
Sure. Here is the link

https://colab.research.google.com/drive/1e8IQIeqOvecb2xu-otMfXMmIgpEDnaKK

**Comment by gowthamkpr** [comment]
Please go through the similar issue [here](https://www.reddit.com/r/tensorflow/comments/bcl5ke/help_with_custom_training_loops_in_tensorflow_20/) and let me know if it helps @Dieblitzen

**Comment by Dieblitzen** [comment]
I saw that issue earlier, but I find it odd that no matter how many times I run the code (in the colab link) it always emits that warning. How can the gradients systematically always be zero for those conv_transpose bias and kernel terms, no matter how many times I run it?

**Comment by gowthamkpr** [comment]
Please look at my gist where the warning is no longer present [here](https://colab.sandbox.google.com/gist/gowthamkpr/733310a5fe0e35c52ffde2a5e869355a/copy-of-tf2-0_convtranspose_issue.ipynb)

**Comment by Dieblitzen** [comment]
@gowthamkpr Yes that surpasses the warning, but my question is why are the gradients for those bias/weight terms always 0? Even if I manually set the gradients to 0 using the code you shared, I still find that the gradients are 0 for every run. Do you know why this is happening?

**Comment by tanzhenyu** [comment]
@Dieblitzen Your code has a bug, it should be:
```python
if up_sampled.shape[1:3] != largest_res.shape[1:3]:
        up_sampled = tf.image.resize(up_sampled, size=largest_res.shape[1:3])
```

not:
```python
if up_sampled.shape[1:3] != largest_res.shape[1:3]:
        up_sampled = tf.image.resize(t, size=largest_res.shape[1:3])
```
